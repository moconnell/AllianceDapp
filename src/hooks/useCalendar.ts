import { useCallback, useEffect, useState } from "react";

import { useWeb3Context } from "../context/Web3Context";
import assert from "assert";
import {
  Calendar,
  CalendarFactory,
  CalendarFactory__factory,
  Calendar__factory,
} from "../typechain-types";
import DaysOfWeek from "../types/daysOfWeek";
import {
  assertValid,
  compare,
  fromTotalMins,
  totalMinutes,
} from "../utils/timeUtils";
import Meeting from "../types/meeting";
import ProfileInfo from "../types/profileInfo";
import AvailabilityInfo from "../types/availabilityInfo";
import { DateTime } from "luxon";

const NULL_ADDRESS = "0x0000000000000000000000000000000000000000";

const toString = (value: any) => (value ? String(value) : "");

export const useCalendar = () => {
  const { address, signer, web3Provider } = useWeb3Context();
  const [calendarFactory, setCalendarFactory] = useState<CalendarFactory>();
  const [calendar, setCalendar] = useState<Calendar>();
  const [availability, setAvailability] = useState<AvailabilityInfo>();
  const [profile, setProfile] = useState<ProfileInfo>();

  const { REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS } = process.env;
  assert(
    REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS,
    "REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS not set"
  );

  useEffect(() => {
    function getCalendarFactory() {
      if (!web3Provider || !address || !signer) return undefined;
      return CalendarFactory__factory.connect(
        REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS!,
        signer
      );
    }

    const fac = getCalendarFactory();
    if (fac) setCalendarFactory(fac);
  }, [REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS, address, signer, web3Provider]);

  useEffect(() => {
    async function getCalendar() {
      if (!address || !calendarFactory || !web3Provider) return;

      const calendarAddress = await calendarFactory.userToCalendar(address);
      if (calendarAddress === NULL_ADDRESS) {
        console.log(`wallet ${address} has no saved calendar`);
        return;
      }

      console.log(`fetching calendar at ${calendarAddress}`);

      const cal = Calendar__factory.connect(calendarAddress, web3Provider);
      setCalendar(cal);
    }

    getCalendar();
  }, [address, calendarFactory, web3Provider]);

  useEffect(() => {
    async function loadProfileAvailability() {
      if (!calendar) return;

      const newProfile = await calendar.profile();
      setProfile(newProfile);

      let newAvailability = await calendar.availability();
      const from = fromTotalMins(newAvailability.earliestStartMinutes);
      const to = fromTotalMins(
        newAvailability.earliestStartMinutes + newAvailability.minutesAvailable
      );
      setAvailability({ ...newAvailability, from, to });
    }

    loadProfileAvailability();
  }, [calendar]);

  useEffect(() => {
    const createCalendar = async () => {
      if (
        calendar ||
        !calendarFactory ||
        !web3Provider ||
        !profile ||
        !availability
      )
        return;

      assert(profile.email, "No email");
      assertValid(availability.from);
      assertValid(availability.to);
      assert(
        compare(availability.from!, availability.to!) < 0,
        "from after to"
      );

      const p = {
        description: toString(profile.description),
        email: profile.email!,
        picture: toString(profile.picture),
        url: toString(profile.url),
        username: toString(profile.username),
      };

      const a = {
        availableDays: availability.availableDays ?? DaysOfWeek.None,
        earliestStartMinutes: totalMinutes(availability.from!),
        minutesAvailable:
          totalMinutes(availability.to!) - totalMinutes(availability.from!),
        location: availability.location!,
        timeZone: availability.timeZone!,
      };

      console.log("creating calendar now ... ", p, a);

      const tx = await calendarFactory.createCalendar(p, a);
      const receipt = await tx.wait();
      const txEvent = receipt.events?.[0]?.args;
      const calendarAddr = txEvent?.calendar;
      const cal = Calendar__factory.connect(calendarAddr, web3Provider);

      console.log("calendar created ok", cal);

      setCalendar(cal);
    };

    createCalendar();
  }, [
    profile,
    availability,
    calendar,
    calendarFactory,
    web3Provider,
    address,
    REACT_APP_WEB3_CONTRACT_FACTORY_ADDRESS,
  ]);

  const getMeetings = useCallback(
    async (year: number, month: number, date: number) => {
      if (!calendar) return undefined;

      console.log(`getMeetings(year: ${year}, month: ${month}, date: ${date})`);

      const meetingsOnDate = await calendar.getMeetings(year, month, date);

      return meetingsOnDate.map((m) => {
        let from = new Date(date);
        from.setMinutes(m.startMinutes);

        let to = new Date(from);
        to.setMinutes(to.getMinutes() + m.durationMinutes);

        return {
          date: from,
          endDate: to,
          attendee: m.attendee,
          description: `${m.durationMinutes} minute meeting`,
        } as Meeting;
      });
    },
    [calendar]
  );

  const getProfileAvailability = useCallback(
    async (calendarAddress: string) => {
      if (!web3Provider) return undefined;

      console.log(
        `getProfileAvailability(calendarAddress: ${calendarAddress})`
      );

      const otherCalendar = Calendar__factory.connect(
        calendarAddress,
        web3Provider
      );

      const pa: [ProfileInfo, AvailabilityInfo] = [
        await otherCalendar.profile(),
        await otherCalendar.availability(),
      ];

      return pa;
    },
    [web3Provider]
  );

  const getAvailableTimes = useCallback(
    async (calendarAddress: string, date: Date, durationMinutes: number) => {
      if (!web3Provider) return undefined;

      console.log(
        `getAvailableTimes(calendarAddress: ${calendarAddress}, date: ${date}, duration: ${durationMinutes})`
      );

      const otherCalendar = Calendar__factory.connect(
        calendarAddress,
        web3Provider
      );

      const year = date.getFullYear();
      const month = date.getMonth() + 1;
      const day = date.getDate();
      const { times, timeZone } = await otherCalendar.getAvailableTimes(
        year,
        month,
        day,
        durationMinutes
      );

      return times
        .filter((startMinutes) => startMinutes > 0)
        .map((startMinutes) => {
          const hour = Math.floor(startMinutes / 60),
            minute = startMinutes % 60;

          return DateTime.fromObject(
            { year, month, day, hour, minute },
            { zone: timeZone }
          );
        });
    },
    [web3Provider]
  );

  const setProfileAvailability = (info: ProfileInfo & AvailabilityInfo) => {
    setProfile(info);
    setAvailability(info);
  };

  return {
    availability,
    calendar,
    getAvailableTimes,
    getMeetings,
    getProfileAvailability,
    profile,
    setProfileAvailability,
  };
};
