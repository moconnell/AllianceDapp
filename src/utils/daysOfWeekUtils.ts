import DayOfWeek from "../types/dayOfWeek";
import DaysOfWeek from "../types/daysOfWeek";

export const toDaysOfWeek = (day: DayOfWeek) => DaysOfWeek[DayOfWeek[day] as keyof typeof DaysOfWeek];