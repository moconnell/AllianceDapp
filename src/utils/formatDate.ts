export function getMeridian(hour, minutes) {
  const formattedMinutes = minutes > 9 ? minutes : `0${minutes}`;
  return hour > 11
    ? `${hour - 12 || 12}:${formattedMinutes}PM`
    : `${hour}:${formattedMinutes}AM`;
}
