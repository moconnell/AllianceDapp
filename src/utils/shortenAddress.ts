export function shortenAddress(address) {
  const length = address.length;
  return `${address.slice(0, 5)}...${address.slice(length - 4, length - 1)}`;
}
