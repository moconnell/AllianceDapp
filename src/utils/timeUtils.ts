import Time from "../types/time";
import assert from "assert";

export const totalMinutes = (t: Time) => t.hours * 60 + (t.minutes ?? 0);

export const totalSeconds = (t: Time) =>
  t.hours * 60 * 60 + (t.minutes ?? 0) * 60 + (t.seconds ?? 0);

export const fromTotalMins = (mins: number) => {
  const hours = Math.floor(mins / 60) % 24;
  const minutes = mins % 60;
  assert(hours < 24);
  assert(minutes < 60);
  return { hours, minutes } as Time;
};

export const assertValid = (t?: Time) => {
  assert(t);
  assert(t.hours >= 0 && t.hours < 24);
  assert(t.minutes >= 0 && t.minutes < 60);
};

export const compare = (t1: Time, t2: Time) => {
  const s1 = totalSeconds(t1);
  const s2 = totalSeconds(t2);
  if (s1 === s2) return 0;
  if (s1 < s2) return -1;
  return 1;
};
