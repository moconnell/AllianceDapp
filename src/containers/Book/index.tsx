import {
  Button,
  Container,
  Flex,
  Heading,
  Table,
  Tbody,
  Td,
  Th,
  Tr,
} from "@chakra-ui/react";
import { DateTime } from "luxon";
import { useEffect, useState } from "react";
import Calendar from "react-calendar";
import { useParams } from "react-router-dom";
import { useCalendar } from "../../hooks";
import DaysOfWeek from "../../types/daysOfWeek";
import { tomorrow } from "../../utils/timeUtils";
import ModalComponent from "../../components/Modal";
import LoadingTransaction from "../../components/Modal/components/LoadingTransaction";
import TimeList from "../../components/TimeList";

const Book = () => {
  const { calendarAddress } = useParams();
  const { getProfileAvailability, getAvailableTimes } = useCalendar();
  const [selectedTime, setselectedTime] = useState<DateTime | undefined>();
  const [availableDays, setAvailableDays] = useState(DaysOfWeek.None);
  const [availableTimes, setAvailableTimes] = useState([] as DateTime[]);
  const [description, setDescription] = useState("");
  const [location, setLocation] = useState("");
  const [timeZone, setTimeZone] = useState("");
  const [username, setUsername] = useState("");
  const [showModal, setShowModal] = useState(false);
  const [date, setDate] = useState(tomorrow());
  const [durationMinutes] = useState(60);
  const now = new Date();

  useEffect(() => {
    const loadProfileAvailability = async () => {
      if (!calendarAddress) return;
      const pa = await getProfileAvailability(calendarAddress);
      if (pa) {
        const [
          { username, description },
          { availableDays, location, timeZone },
        ] = pa;
        if (availableDays) setAvailableDays(availableDays);
        if (location) setLocation(location);
        if (timeZone) setTimeZone(timeZone);
        if (username) setUsername(username);
        if (description) setDescription(description);
      }
    };

    loadProfileAvailability();
  }, [calendarAddress, getProfileAvailability]);

  useEffect(() => {
    const loadTimes = async () => {
      if (!calendarAddress) return;
      const availableTimes_ = await getAvailableTimes(
        calendarAddress,
        date,
        durationMinutes
      );
      if (availableTimes_) {
        console.log(availableTimes_);
        setAvailableTimes(availableTimes_);
      }
    };

    loadTimes();
  }, [calendarAddress, date, durationMinutes, getAvailableTimes]);

  function handleCloseModal() {
    setShowModal(false);
  }

  return (
    <Container maxW="container.xl" p={1}>
      <Heading color="raid.100" fontFamily="Inter">
        Book a {durationMinutes} minute meeting{username && ` with ${username}`}
        ...
      </Heading>
      <Flex
        h={{ base: "auto", md: "100%" }}
        py={10}
        direction={{ base: "column", md: "row" }}
      >
        {(location || timeZone || description) && (
          <Container paddingTop="24">
            <Table>
              <Tbody fontFamily="Inter">
                {location && (
                  <Tr>
                    <Th border={0} color="gray.500">
                      Location
                    </Th>
                    <Td border={0} color="gray.300" fontSize="sm">
                      {location}
                    </Td>
                  </Tr>
                )}
                {timeZone && (
                  <Tr>
                    <Th border={0} color="gray.500">
                      Time-Zone
                    </Th>
                    <Td border={0} color="gray.300" fontSize="sm">
                      {timeZone}
                    </Td>
                  </Tr>
                )}
                {description && (
                  <Tr>
                    <Th border={0} color="gray.500">
                      About
                    </Th>
                    <Td border={0} color="gray.300" fontSize="sm">
                      {description}
                    </Td>
                  </Tr>
                )}
              </Tbody>
            </Table>
          </Container>
        )}
        <Container p={0}>
          <Calendar
            onChange={setDate}
            defaultValue={date}
            tileDisabled={({ date }) =>
              date < now ||
              ((1 << date.getDay()) & availableDays) !== 1 << date.getDay()
            }
          />
        </Container>
        <Container p={4}>
          <TimeList
            selectedTime={selectedTime}
            onChange={(time: any) => {
              setselectedTime(time);
            }}
            times={availableTimes}
            duration={durationMinutes}
          />
        </Container>
      </Flex>
      <Container pr={10} maxW="xs">
        <Button onClick={() => setShowModal(true)}>Book meeting</Button>
      </Container>
      <ModalComponent showModal={showModal} closeModal={handleCloseModal}>
        <LoadingTransaction />
      </ModalComponent>
    </Container>
  );
};

export default Book;
