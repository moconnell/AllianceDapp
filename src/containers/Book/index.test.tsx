import { render, waitFor } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { DateTime } from "luxon";
import { Route, Router, Routes } from "react-router-dom";
import DaysOfWeek from "../../types/daysOfWeek";
import { range } from "../../utils/range";
import { useCalendar } from "../../hooks";
import AvailabilityInfo from "../../types/availabilityInfo";
import ProfileInfo from "../../types/profileInfo";
import Book from "./index";

jest.mock("../../hooks");
jest.mock("../../components/Calendar", () => () => "((Calendar))");
jest.mock("../../components/Modal", () => () => "((ModalComponent))");
jest.mock(
  "../../components/Modal/components/LoadingTransaction",
  () => () => "((LoadingTransaction))"
);
jest.mock("../../components/TimeList", () => () => "((TimeList))");

describe("<Book />", () => {
  const mockGetProfileAvailability = jest.fn();
  const mockGetAvailableTimes = jest.fn();

  afterEach(() => jest.resetAllMocks());
  afterAll(() => jest.restoreAllMocks());

  it("should render correctly", async () => {
    (useCalendar as jest.Mock).mockImplementation(() => {
      return {
        getProfileAvailability: mockGetProfileAvailability,
        getAvailableTimes: mockGetAvailableTimes,
      };
    });

    mockGetProfileAvailability.mockImplementation(() => {
      const timeZone = "Europe/London";
      const location = timeZone.split("/")[1];
      const profileAvailability: [ProfileInfo, AvailabilityInfo] = [
        { username: "mr_bojangles", description: "legend" },
        {
          location,
          timeZone,
          availableDays: DaysOfWeek.MonFri,
        },
      ];

      return Promise.resolve(profileAvailability);
    });

    mockGetAvailableTimes.mockResolvedValue(
      range(9, 14).map((hour) => DateTime.fromObject({ hour }))
    );

    const history = createMemoryHistory();
    history.push("/book/0x23dB4a08f2272df049a4932a4Cc3A6Dc1002B33E");

    const { asFragment, getByTestId } = render(
      <Router location={history.location} navigator={history}>
        <Routes>
          <Route path="/book/:calendarAddress" element={<Book />} />
        </Routes>
      </Router>
    );

    await waitFor(() =>
      expect(getByTestId("container:profile")).toBeInTheDocument()
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
