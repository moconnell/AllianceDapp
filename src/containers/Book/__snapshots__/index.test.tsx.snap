// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Book /> should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-container css-122pq3t"
    data-testid="container:book"
  >
    <h2
      class="chakra-heading css-147w9li"
    >
      Book a 60 minute meeting with mr_bojangles...
    </h2>
    <div
      class="css-b82b48"
    >
      <div
        class="chakra-container css-19n8dai"
        data-testid="container:profile"
      >
        <table
          class="chakra-table css-0"
          role="table"
        >
          <tbody
            class="css-13c8hy8"
          >
            <tr
              class="css-0"
              role="row"
            >
              <th
                class="css-1u3nbe2"
              >
                Location
              </th>
              <td
                class="css-17arw3k"
                role="gridcell"
              >
                London
              </td>
            </tr>
            <tr
              class="css-0"
              role="row"
            >
              <th
                class="css-1u3nbe2"
              >
                Time-Zone
              </th>
              <td
                class="css-17arw3k"
                role="gridcell"
              >
                Europe/London
              </td>
            </tr>
            <tr
              class="css-0"
              role="row"
            >
              <th
                class="css-1u3nbe2"
              >
                About
              </th>
              <td
                class="css-17arw3k"
                role="gridcell"
              >
                legend
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="chakra-container css-old1by"
      >
        ((Calendar))
      </div>
      <div
        class="chakra-container css-2rlxtj"
      >
        ((TimeList))
      </div>
    </div>
    <div
      class="chakra-container css-5exlul"
    >
      <button
        class="chakra-button css-5c2176"
        type="button"
      >
        Book meeting
      </button>
    </div>
    ((ModalComponent))
  </div>
</DocumentFragment>
`;
